### PR TITLE
The great toxification (part 3)

### PR DESCRIPTION
**Pull request by stephenfin**
Subject: Centralize configuration

Centralize as much configuration as possible inside `setup.cfg` and `tox.ini`, to avoid polluting the top level directory

### Feature or Bugfix
- Feature

### Purpose
To paraphrase the Zen of Python (import this):

> tox is one honking great idea -- let's use more of it!

Start modernizing the test infrastructure that's using make and other hand-crafted tools by replacing them with tox-based derivatives.

**This is part twp.**

### Detail
This is a multi-step process, with the end goal of completely removing the Makefile and running everything through `tox`. The first fours steps of these can be done in parallel, but combined they block the final two.

1. Make `tox` a little more friendly by enabling extras like coverage, test profiling etc. by default, and closing gaps with the current Makefile targets such as PyLint and building docs in particular formats
2. Replace/remove custom tooling found in `utils` in favor of modern alternatives like `flake8`
3. Centralize as much configuration as possible inside `setup.cfg` and `tox.ini`, to avoid polluting the top level directory (**this change**)
4. Clean up requirements (`test-reqs.txt`, `setup.py`) to ensure virtualenv and non-virtualenv builds are identical

The remaining two steps are blocked by the above:

5. Switch CIs to use `tox` instead of the Makefile, and update `CONTRIBUTING.rst` to refer to tox. Deprecate the Makefile
6. (Bonus) Integrate [codecov.io](https://codecov.io) to make use of the coverage stats we now have

### Relates
- N/A

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -28,5 +28,18 @@ max-line-length = 95
 ignore = E116,E241,E251
 exclude = .git,.tox,.venv,tests/*,build/*,sphinx/search/*,sphinx/pycode/pgen2/*,doc/ext/example*.py
 
+[mypy]
+python_version = 2.7
+show_column_numbers = True
+show_error_context = False
+ignore_missing_imports = True
+follow_imports = skip
+cache_dir = /dev/null
```
Comment (tk0miya) [inline setup.cfg]: I feel this is harmful. why do you disable caches?

File: setup.cfg [hunk 2]
```diff
@@ -28,5 +28,18 @@ max-line-length = 95
...
+incremental = True
+check_untyped_defs = True
+warn_no_return = True
+warn_redundant_casts = False
```
Comment (tk0miya) [inline setup.cfg]: It seems some configurations are added silently. Why are they needed? please let us know.

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 1 above)
Comment (tk0miya) [reply setup.cfg]: In addition, I think this brings troubles to windows environment.

Comment (stephenfin) [reply setup.cfg]: > I feel this is harmful. why do you disable caches?

This was causing me issues when working on #4097. Whenever I added a line like [this](https://github.com/sphinx-doc/sphinx/pull/4097/commits/3559577d5ef37211a1e3cd9222d560a0541d6328#diff-c3a9ba1cd0902193425499e0cf7827b6R288) I would get error messages. All this does is improve performance a small bit, so I figured I could accept the cost. However...

> In addition, I think this brings troubles to windows environment.

That's a good point. I guess I'll have to drop this bit.

File: setup.cfg (hunk 2 above)
Comment (stephenfin) [reply setup.cfg]: Oops - these were experiments. I'll remove them

**Comment by stephenfin** [comment]
I've removed the erroneous entries to the `[mypy]` section and expanded the commit message for the two I actually want

**Review by tk0miya — APPROVED** [review]
LGTM, it's reasonable!

**Comment by tk0miya** [comment]
Just merged!

**Pull request merged**